File: sdks/php/src/Api/Contacts.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\AddToSegmentBody;
use Apollo\Deploy\Signal\Sdk\ContactPageResponse;
use Apollo\Deploy\Signal\Sdk\ContactResponse;
use Apollo\Deploy\Signal\Sdk\CreateContactBody;
use Apollo\Deploy\Signal\Sdk\EmailValidationStatusResponse;
use Apollo\Deploy\Signal\Sdk\EngagementScoreResponse;
use Apollo\Deploy\Signal\Sdk\RecordValidationBody;
use Apollo\Deploy\Signal\Sdk\SetImageUrlBody;
use Apollo\Deploy\Signal\Sdk\UpdateContactBody;
use Apollo\Deploy\Signal\Sdk\UpdateTopicsBody;
use Psr\Http\Message\ResponseInterface;

/** ContactsAPI domain operations. */
class Contacts
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return ContactPageResponse
     * @throws SDKError
     */
    public function listContacts(
        string $projectId,
    ): ContactPageResponse
    {
        $path = '/v1/projects/${projectId}/contacts';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactPageResponse::class, 'fromArray')) {
            return ContactPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return ContactResponse
     * @throws SDKError
     */
    public function getContact(
        string $projectId,
        string $contactId,
    ): ContactResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactResponse::class, 'fromArray')) {
            return ContactResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return array
     * @throws SDKError
     */
    public function listContactSegments(
        string $projectId,
        string $contactId,
    ): array
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/segments';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        return is_array($data) ? $data : [];
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return array
     * @throws SDKError
     */
    public function getContactTopics(
        string $projectId,
        string $contactId,
    ): array
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/topics';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        return is_array($data) ? $data : [];
    }

    /**
     * @param string $projectId
     * @param CreateContactBody $input
     * @return ContactResponse
     * @throws SDKError
     */
    public function createContact(
        string $projectId,
        CreateContactBody $input,
    ): ContactResponse
    {
        $path = '/v1/projects/${projectId}/contacts';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactResponse::class, 'fromArray')) {
            return ContactResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @param UpdateContactBody $input
     * @return ContactResponse
     * @throws SDKError
     */
    public function updateContact(
        string $projectId,
        string $contactId,
        UpdateContactBody $input,
    ): ContactResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PATCH',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactResponse::class, 'fromArray')) {
            return ContactResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return void
     * @throws SDKError
     */
    public function deleteContact(
        string $projectId,
        string $contactId,
    ): void
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return ContactResponse
     * @throws SDKError
     */
    public function uploadContactImage(
        string $projectId,
        string $contactId,
    ): ContactResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/image';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactResponse::class, 'fromArray')) {
            return ContactResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @param SetImageUrlBody $input
     * @return ContactResponse
     * @throws SDKError
     */
    public function setContactImageUrl(
        string $projectId,
        string $contactId,
        SetImageUrlBody $input,
    ): ContactResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/image';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PUT',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(ContactResponse::class, 'fromArray')) {
            return ContactResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return void
     * @throws SDKError
     */
    public function deleteContactImage(
        string $projectId,
        string $contactId,
    ): void
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/image';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @param AddToSegmentBody $input
     * @return void
     * @throws SDKError
     */
    public function addContactToSegment(
        string $projectId,
        string $contactId,
        AddToSegmentBody $input,
    ): void
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/segments';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @param string $segmentId
     * @return void
     * @throws SDKError
     */
    public function removeContactFromSegment(
        string $projectId,
        string $contactId,
        string $segmentId,
    ): void
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/segments/${segmentId}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);
        $path = str_replace('${segmentId}', rawurlencode(Transport::wireValue($segmentId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @param UpdateTopicsBody $input
     * @return void
     * @throws SDKError
     */
    public function updateContactTopics(
        string $projectId,
        string $contactId,
        UpdateTopicsBody $input,
    ): void
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/topics';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'PATCH',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return array
     * @throws SDKError
     */
    public function getContactActivity(
        string $projectId,
        string $contactId,
    ): array
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/activity';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        return is_array($data) ? $data : [];
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return EngagementScoreResponse
     * @throws SDKError
     */
    public function getContactEngagementScore(
        string $projectId,
        string $contactId,
    ): EngagementScoreResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/engagement';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(EngagementScoreResponse::class, 'fromArray')) {
            return EngagementScoreResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return EmailValidationStatusResponse
     * @throws SDKError
     */
    public function getContactEmailValidationStatus(
        string $projectId,
        string $contactId,
    ): EmailValidationStatusResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/email-validation';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(EmailValidationStatusResponse::class, 'fromArray')) {
            return EmailValidationStatusResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @return EngagementScoreResponse
     * @throws SDKError
     */
    public function refreshContactEngagementScore(
        string $projectId,
        string $contactId,
    ): EngagementScoreResponse
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/engagement/refresh';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(EngagementScoreResponse::class, 'fromArray')) {
            return EngagementScoreResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $contactId
     * @param RecordValidationBody $input
     * @return void
     * @throws SDKError
     */
    public function recordContactEmailValidation(
        string $projectId,
        string $contactId,
        RecordValidationBody $input,
    ): void
    {
        $path = '/v1/projects/${projectId}/contacts/${contactId}/email-validation';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${contactId}', rawurlencode(Transport::wireValue($contactId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/contacts.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ AddToSegmentBody, ContactPageResponse, ContactResponse, CreateContactBody, EmailValidationStatusResponse, EngagementScoreResponse, GetContactActivityResponse, GetContactTopicsResponse, ListContactSegmentsResponse, RecordValidationBody, SetImageUrlBody, UpdateContactBody, UpdateTopicsBody };

/// ContactsAPI API operations.
pub struct ContactsAPI {
    transport: Transport,
}

impl ContactsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_contacts(
        &self,
        project_id: String,
    ) -> Result<ContactPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_contact(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<ContactResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn list_contact_segments(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<ListContactSegmentsResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/segments".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ListContactSegmentsResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_contact_topics(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<GetContactTopicsResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/topics".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<GetContactTopicsResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn create_contact(
        &self,
        project_id: String,
        body: &CreateContactBody,
    ) -> Result<ContactResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn update_contact(
        &self,
        project_id: String,
        contact_id: String,
        body: &UpdateContactBody,
    ) -> Result<ContactResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactResponse>(
            Method::PATCH,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn delete_contact(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn upload_contact_image(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<ContactResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/image".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactResponse>(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn set_contact_image_url(
        &self,
        project_id: String,
        contact_id: String,
        body: &SetImageUrlBody,
    ) -> Result<ContactResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/image".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<ContactResponse>(
            Method::PUT,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn delete_contact_image(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/image".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn add_contact_to_segment(
        &self,
        project_id: String,
        contact_id: String,
        body: &AddToSegmentBody,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/segments".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn remove_contact_from_segment(
        &self,
        project_id: String,
        contact_id: String,
        segment_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/segments/${segmentId}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));
        path = path.replace("${segmentId}", &encode_path(&segment_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn update_contact_topics(
        &self,
        project_id: String,
        contact_id: String,
        body: &UpdateTopicsBody,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/topics".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::PATCH,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn get_contact_activity(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<GetContactActivityResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/activity".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<GetContactActivityResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_contact_engagement_score(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<EngagementScoreResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/engagement".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<EngagementScoreResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn get_contact_email_validation_status(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<EmailValidationStatusResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/email-validation".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<EmailValidationStatusResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn refresh_contact_engagement_score(
        &self,
        project_id: String,
        contact_id: String,
    ) -> Result<EngagementScoreResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/engagement/refresh".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<EngagementScoreResponse>(
            Method::POST,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn record_contact_email_validation(
        &self,
        project_id: String,
        contact_id: String,
        body: &RecordValidationBody,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/contacts/${contactId}/email-validation".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${contactId}", &encode_path(&contact_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await.map(|_| ())
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
